proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+frequently updated large collections of resources, e.g. with 1M endpoints updating
+every 10s.  For this situation, we introduce *glob
+collections*, which allow the `<T>Collection` directory to be elided when
+resources in a collection share a directory structure and a `/*` component is appended to
+the resource path. Continuing the previous example, this now looks like:
+
+1. Client requests
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/foo/*.`
+2. Server responds with resources
+   [`xdstp://some-authority/envoy.config.listener.v3.Listener/foo/bar,
+   xdstp://some-authority/envoy.config.listener.v3.Listener/foo/baz].`
+
+Since each resource returned is subject to independent update via delta xDS and
+there is no explicit collection directory to update, glob collections are highly
+scalable.
+
+#### Use cases
+
+This proposal provides multiple collection mechanisms, as there is a need to
+tradeoff ergonomics (i.e. being able to reason about a collection via single
+object), performance (i.e. number of round trips) and scalability (e.e. avoiding
+a collection directory becoming a bottleneck).
+
+Some suggested use patterns for the collections are as follows:
+
+*   Federated collections (list collections with references)
+*   HTTP CDN delivery of configuration (list collections with references)
+*   Handcrafted configuration on a filesystem (list collections with inline entries)
+*   Migration of an existing state-of-the-world (SotW) Envoy configuration (list collections with inline entries)
+*   LEDS and high scalability delta APIs (glob collections)
+
+For any other use case with a:
+
+*   Small collection of objects or relatively static configuration;
+    list collections with inline entries are recommended. This is both
+    performant and requires minimal management server state tracking and round
+    trips.
+*   Medium sized collection of objects; list collections with reference entries.
+    The management server only needs to handle subscribe/unsubscribe state
+    tracking, since both the collection and its constituent resources are
+    independent xDS resources.
+*   For any other use case with a large collection of objects and/or highly
+    dynamic updates, glob collections are recommended. The management server needs to
+    provide full delta diffs, since there is no explicit collection object to
+    inform the client when resources come and go.
+
+### Transport network addressing
+
+An `xdstp://` URI specifies a logic authority and it’s necessary to map to some physical 
+network address during the resource fetch process. In the existing xDS configuration
+fetch configuration, this capability is provided by a `ConfigSource` delivered alongside a
+resource name in the xDS resource graph.
+
+This proposal adds to `ConfigSource` the ability to specify the
+authorities for which the `ConfigSource` may be used:
+
+```
+message Authority {
```

Comment:
For gRPC, I think we're going to want to structure this differently.  We're going to have a map of authority name to the configuration for that authority.  The config for a given authority will include the following:

- An ordered list of (the equivalent of) `ConfigSource`s that indicate which server(s) can provide data for that authority.  The client will prefer servers in the order listed here, which provides a convenient way of configuring failover.
- Node context params will be specified here, since these are likely to be different for different authorities.
- (Not part of this design, but in the follow-up xRFC I plan to write for dynamic context parameter negotiation, this will include setting the name of the DCP resource that will be used for this authority.)

Reply:
What about the `ConfigSource` that is not in the bootstrap but already sitting next to the resource in question, e.g. https://github.com/envoyproxy/envoy/blob/4ecf3f7efed97529dc064223f3697a8d4d1101b9/api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#L528. Should there be some way to allow this to be optionally used as the `ConfigSource` for the authority as a bootstrap override? It seems weird to give up on allowing dynamic config source completely when we have this flexibility today.

Reply:
Friendly ping on this one @markdroth. I'm happy to add a sketch of a bootstrap and authority mapping, but it would be good to describe how this interoperates with existing `ConfigSource` that sit next to resource names.

Reply:
gRPC does not currently support `ConfigSources` that sit next to resource names (which I refer to as "dynamic `ConfigSource`s") that point at arbitrary remote servers, because there are concerns about trusting the management server to tell the client what credentials to use to talk to a remote server.  Currently, our bootstrap file specifies a single ADS server from which we get all of our data, and we require dynamic `ConfigSource`s to populate the `ads` field, so that they always specify resources from the same server.  As part of implementing federation in gRPC, we will change our bootstrap file to support the authority map I described above, and then servers will be able to tell clients to get data from a different server via the authority in the resource URI, as long as that authority is already present in the client's bootstrap file.  But we still don't plan to allow servers to send arbitrary dynamic `ConfigSource`s to clients.  (The only change we will make is that we will start allowing the `self` field, and we will treat `ads` the same as `self`  for backward compatibility, which will take us in the same direction as envoyproxy/envoy#13951.)  So for gRPC, this is not an issue.

That having been said, if Envoy is going to continue to support these dynamic `ConfigSource`s, it seems to me that you can continue to support that regardless of how the per-authority `ConfigSource`s are laid out in your bootstrap file.  I think your original plan was that if there is a dynamic `ConfigSource`, it will be used, and if not, you will look up the authority from the resource name in the bootstrap file to see if there's a `ConfigSource` present for that authority.  I think you can stick with that approach, regardless of what exact format you use for the `ConfigSource`s in the bootstrap file.

I think the main advantages of having an authority map in the boostrap file are that (a) it provides a way to specify the priority order of the list of servers to use for a given authority and (b) it provides a place to put other per-authority configuration, such as cache key constraint configuration.

Reply:
So, if we stick with what is here and either spell out the bootstrap `ConfigSource` style or leave to another PR this is fine?

I'm happy to dump the gRPC opinionated bootstrap schema here and adopt, as I think we want this in Envoy as well. Do you have examples?

Reply:
OK, I'm adding just a simple `repeated` to give an idea of what is possible and will quality the dynamic `ConfigSource` is not necessary if a client has another resolution mechanism.